File: 2021/day-02/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readToArrays(path string) ([]string, []int, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	var cmds []string
	var values []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " ")
		i, err := strconv.Atoi(s[1])
		if err != nil {
			panic(err)
		}
		cmds = append(cmds, s[0])
		values = append(values, i)
	}
	return cmds, values, scanner.Err()
}

func parseCmd(cmd string, value int, depth int, x int, y int) (int, int, int) {
	switch cmd {
	case "forward":
		x += value
		depth += y * value
	case "up":
		y -= value
	case "down":
		y += value
	default:
		panic("unknown command")
	}
	return depth, x, y
}

func main() {
	cmds, values, err := readToArrays("input.txt")
	if err != nil {
		panic(err)
	}

	var depth int
	var x int
	var y int
	for i := 0; i < len(cmds); i++ {
		depth, x, y = parseCmd(cmds[i], values[i], depth, x, y)
	}

	score := x * y
	fmt.Printf("Part 1: position (%d, %d), score=%d\n", x, y, score)

	score = x * depth
	fmt.Printf("Part 2: position (%d, %d, %d), score=%d\n", depth, x, y, score)
}

File: 2021/day-02/main.py
def read_data(filename):
    cmds = []
    with open(filename, 'r') as f:
        for l in f.readlines():
            cmd, value = l.strip().split(' ')
            cmds.append((cmd, int(value)))

    return cmds

def parse_cmd(cmd, value, depth, x, y):
    if cmd == 'forward':
        x += value
        depth += y * value
    elif cmd == 'up':
        y -= value
    elif cmd == 'down':
        y += value
    else:
        raise ValueError(f'Invalid command `{cmd}`')

    return depth, x, y


def main():

    cmds = read_data('input.txt')
    depth, x, y = 0, 0, 0
    for cmd, value in cmds:
        depth, x, y = parse_cmd(cmd, value, depth, x, y)

    print('--- Part 1 ---')
    print(f'Final position: {x}, {y}')
    print(f'Final score: {x * y}')

    print('--- Part 2 ---')
    print(f'Final position: {depth}, {x}, {y}')
    print(f'Final score: {depth * x}')


if __name__ == '__main__':
    main()